optimum/intel/neural_compressor/trainer.py

```diff
@@ -553,17 +553,18 @@ def compute_loss(self, model, inputs, return_outputs=False):
             self.is_in_train
             and hasattr(self, "agent")
             and hasattr(self.agent, "criterion")
-            and hasattr(self.agent.criterion, "teacher_model_forward")
+            and self.agent.criterion is not None
         ):
             logits = self._get_logits(outputs)
-            teacher_outputs = inputs.pop("teacher_outputs", None)
-            # Compute teacher model outputs
-            if teacher_outputs is None:
-                teacher_outputs = self.agent.criterion.teacher_model_forward(
-                    inputs, teacher_model=self.agent.teacher_model._model
-                )
-
-            teacher_logits = self._get_logits(teacher_outputs)
+            teacher_logits = inputs.pop("teacher_logits", None)
+            if hasattr(self.agent, "on_after_compute_loss"):

```

Comment:
Is it equivalent to `hasattr(self.agent.criterion, "teacher_model_forward")` ?

Reply:
on_after_compute_loss is not equivalent to teacher_model_forward, in future , we will use on_after_compute_loss callback to compute distillation loss. the usage is like below:
student_outputs = student_model(input)
student_loss = user_criterion(student_outputs, labels)
total_loss = agent.on_after_compute_loss(input, student_outputs, student_loss, tearch_outputs)
here, teacher_outputs is optional.
The callback of on_after_compute_loss is a new API in 1.13. The purpose of this is to reuse the user's criterion.
But in on_after_compute_loss, it didn't handle a tuple of student_outputs and teacher_outputs. so we didn't use the on_after_compute_loss directly, in the future I will commit another commit to use on_after_compute_loss directly.

Reply:
My question was more about does `hasattr(self.agent, "on_after_compute_loss")` is equivalent to `hasattr(self.agent.criterion, "teacher_model_forward")` as `self.agent.criterion.teacher_model_forward` is called after this condition. Also thanks for your explanation, does it mean that the `teacher_model_forward` method will be deprecated ? I find this method very useful as it gives us more flexibility : and compute the loss ourselves with the trainer `compute_distillation_loss` method

Reply:
The teacher_model_forward will exist in the old version and the new version in neural_compressor,  we can use it anyway. Yes, if you want to use the trainer compute_distillation_loss method, you can use hasattr(self.agent.criterion, "teacher_model_forward") condition, but it is only in the new version that will return outputs. So you need to judge that is a new version neural_compressor first.